[HTML5] Optional GDNative Support

Faless opened the pull request:
After months of works, refactoring, bloody tears, GDNative is finally working for HTML5 exports, with some quirks (see below).
[Here is a working GDNative C++ demo](https://no-war.fales.me/gdnative/gdnative.html) running on the web.

![demo](https://user-images.githubusercontent.com/1687918/101088330-d42bf200-35b3-11eb-9d3e-8331751c4ebf.png)

The demo was built with [this godot-cpp](https://github.com/Faless/godot-cpp/tree/js/gdnative), [this demo projects](https://github.com/Faless/gdnative-demos/tree/js/simple) and [this 3.2 branch](https://github.com/Faless/godot/tree/js/3.x_gdnative). I'll PR them in the respective repositories.

Building a version that support dynamic linking requires a very recent emscripten version, at the time of writing `2.0.10` has been tested.
You can build an HTML5 template with GDNative support via:
```
scons p=javascript gdnative_enabled=yes
```
(yes, this kinda clash with `module_gdnative_enabled=yes`, should I rename it to `dlopen_enabled`? `dynamic_linking_enabled`?)

In this PR:
- ~ENet patch to workaround emscripten 2.0.9 issue.~ Removed, no longer needed
  ~(now fixed upstream, new version will hopefully be good, once released I'll remove this commit and the draft status).~
- Add function signatures to JavaScript libraries (which is supposed to be required in newer emscripten).
- Add a new "Export Type" section that allows the user to select either Regular, Threads, GDNative build.
  Expected template names are: `webassembly[_$FEATURE]_$TARGET].zip`, e.g `webassembly_gdnative_release.zip`.
  When using custom export that option is only relevant for the GDNative build (otherwise the libraries won't be exported, but of course the selected template must be of the proper type).

![gdnative](https://user-images.githubusercontent.com/1687918/101087743-f2ddb900-35b2-11eb-8897-18f82d72d697.png)

Faless committed:
[HTML5] GDNative support via SIDE_MODULE.

Working with emscripten >= 2.0.10

Faless committed:
[HTML5] Add function signatures to JS libraries.

Faless committed:
[HTML5] Allow selecting the export type.

Available types:
- Regular
- GDNative (support dynamic linking and thus GDNative WASM files)
- Threads (uses WebAssembly Threads)

Faless committed:
[HTML5] Make GDNative support feature-based.

This is suboptimal as it requires adding an extra compile flag, but
rewriting how feature tags work is beyond the scope of this work.

DarkMessiah commented:
Which browser is required?
![image](https://user-images.githubusercontent.com/1756388/101122957-45a48880-3626-11eb-8cc7-c26b118a0077.png)

Faless commented:
> Which browser is required?

Firefox, but the latest version of chrome should also work.

akien-mga commented on modules/gdnative/gdnative_library_editor_plugin.cpp:
```diff
@@ -308,11 +308,11 @@ GDNativeLibraryEditor::GDNativeLibraryEditor() {
-		//		platforms["Javascript"] = platform_html5;
+		NativePlatformConfig platform_html5;
+		platform_html5.name = "HTML5";
+		platform_html5.library_extension = "*.wasm";
+		platform_html5.entries.push_back("wasm");
```
Nitpick but all other definitions have `entries` first and `library_extension` second :P

akien-mga commented on modules/webrtc/library_godot_webrtc.js:
```diff
@@ -109,6 +110,7 @@ const GodotRTCDataChannel = {
 		}
 	},
 
+	godot_js_rtc_datachannel_send__sig: 'iiiii',
```
That's actually `v` in Roman numerals ;)

akien-mga commented on platform/javascript/detect.py:
```diff
@@ -146,14 +148,15 @@ def configure(env):
     # Reduce code size by generating less support code (e.g. skip NodeJS support).
     env.Append(LINKFLAGS=["-s", "ENVIRONMENT=web,worker"])
 
-    # We use IDBFS in javascript_main.cpp. Since Emscripten 1.39.1 it needs to
-    # be linked explicitly.
-    env.Append(LIBS=["idbfs.js"])
```
This is no longer needed to build with early `1.39.x` releases? We use 1.39.9 for the 3.2.4.

Faless replied:
No, it's still needed, but I moved it in the `SCsub` because it must only be linked to the `MAIN_MODULE` (the main runtime program)

akien-mga commented on platform/javascript/export/export.cpp:
```diff
@@ -85,36 +85,44 @@ class EditorHTTPServer : public Reference {
-		String filepath = EditorSettings::get_singleton()->get_cache_dir().plus_file("tmp_js_export");
+		const String cache_path = EditorSettings::get_singleton()->get_cache_dir();
 		const String basereq = "/tmp_js_export";
+		String filepath = "";
```
For the reference `String` is value-initialized to `""` already so the assignment here is not necessary (but doesn't hurt either if it's for readability).

akien-mga commented on platform/javascript/export/export.cpp:
```diff
@@ -213,6 +213,32 @@ class EditorExportPlatformJavaScript : public EditorExportPlatform {
+		EXPORT_MODE_GDNATIVE = 2,
+	};
+
+	String _get_template_name(ExportMode p_mode, bool p_debug) const {
```
I'll get started on updating my build scripts to build all those flavours :)

akien-mga replied:
Though for the `3.2` branch it's going to be a bit tricky as the current Mono version goes with Emscripten 1.39.7 (patched to 1.39.9 because 1.39.7 is broken), which doesn't work for GDNative.

We could possibly make the standard builds with Emscripten 2.0.10+ (normal, threads, gdnative) and Mono builds with Emscripten 1.39.9 (normal, threads). Or work our way through upstream Mono patches to up support to Emscripten 2.0.10 in the Mono 6.12 branch.

akien-mga commented on platform/javascript/os_javascript.cpp:
```diff
@@ -134,6 +134,14 @@ bool OS_JavaScript::_check_internal_feature_support(const String &p_feature) {
 		return true;
 #endif
+#ifndef NO_THREADS
+	if (p_feature == "threads")
```
These should use braces. (Same with the existing `JavaScript` check, it was likely missed by `clang-tidy` as this file isn't compiled on Linux.)

akien-mga commented:
Looks great on a first review :+1:

Faless commented:
@DarkMessiah thanks for the report, I fixed the Chrome issue and uploaded a new version.
It should work for you now.
Please make sure to clear your browser cache and reloading the page.
(I will fix the cache issues with exported games soon, I promise)

Faless committed:
[HTML5] EditorRunNative works with GDNative.

This "breaks" our loading bar logic (libraries are not counted).
Fixing it is non trivial and probably deserves investigating a different
strategy.

Faless commented:
[Emscripten 2.0.10](https://github.com/emscripten-core/emscripten/releases/tag/2.0.10) is out, I removed the ENet patch, and renamed the feature tag `wasm` to `wasm32` (since `wasm64` will be a thing).
This is ready for review.

akien-mga commented:
> (since wasm64 will be a thing).

Exciting!

akien-mga commented:
Thanks!

akien-mga commented:
Should we add a version check in `detect.py` to abort `gdnative_enabled=yes` build if emcc < 2.0.10?

Faless commented:
> Should we add a version check in `detect.py` to abort `gdnative_enabled=yes` build if emcc < 2.0.10?

I'll look into that
